kratos/geometries/coupling_master_slave.h

```diff
@@ -0,0 +1,452 @@
+
+// System includes
+
+// External includes
+
+// Project includes
+
+
+namespace Kratos
+{
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/**
+ * @class CouplingMasterSlave
+ * @ingroup KratosCore
+ * @brief The CouplingMasterSlave acts as topology for faces having a pointer
+ *        to a surface geometry and a list of BrepEdge s. Those edges
+ *        define the trimming of the respective surface.
+ */
+template<class TPointType> class CouplingMasterSlave
+    : public Geometry<TPointType>
+{
+public:
+    ///@}
+    ///@name Type Definitions
+    ///@{
+
+    /**
+     * Geometry as base class.
+     */
+    typedef Geometry<TPointType> BaseType;
+    typedef Geometry<TPointType> GeometryType;
+
+    /**
+     * Pointer definition of CouplingMasterSlave
+     */
+    KRATOS_CLASS_POINTER_DEFINITION( CouplingMasterSlave );
+
+    typedef TPointType PointType;
+
+    typedef GeometryData::IntegrationMethod IntegrationMethod;
+
+    typedef typename BaseType::GeometriesArrayType GeometriesArrayType;
+    typedef typename BaseType::IndexType IndexType;
+    typedef typename BaseType::SizeType SizeType;
+    typedef typename BaseType::PointsArrayType PointsArrayType;
+    typedef typename BaseType::CoordinatesArrayType CoordinatesArrayType;
+
+    typedef typename BaseType::IntegrationPointType IntegrationPointType;
+    typedef typename BaseType::IntegrationPointsArrayType IntegrationPointsArrayType;
+    typedef typename BaseType::IntegrationPointsContainerType IntegrationPointsContainerType;
+
+    typedef typename BaseType::ShapeFunctionsValuesContainerType ShapeFunctionsValuesContainerType;
+    typedef typename BaseType::ShapeFunctionsLocalGradientsContainerType ShapeFunctionsLocalGradientsContainerType;
+    typedef typename BaseType::ShapeFunctionsGradientsType ShapeFunctionsGradientsType;
+    typedef typename BaseType::ShapeFunctionsSecondDerivativesType ShapeFunctionsSecondDerivativesType;
+    typedef typename BaseType::ShapeFunctionsThirdDerivativesType ShapeFunctionsThirdDerivativesType;
+
+    typedef typename BaseType::JacobiansType JacobiansType;
+    typedef typename BaseType::NormalType NormalType;
+
+    // New Shape function typedefs
+    typedef GeometryData::ShapeFunctionsDerivativesType ShapeFunctionsDerivativesType;
+    typedef GeometryData::ShapeFunctionsDerivativesIntegrationPointsType ShapeFunctionsDerivativesIntegrationPointsType;
+    typedef GeometryData::ShapeFunctionsDerivativesVectorType ShapeFunctionsDerivativesVectorType;
+
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    CouplingMasterSlave(
+        typename GeometryType::Pointer pMasterGeometry,
+        typename GeometryType::Pointer pSlaveGeometry,
+        IsGeometryDataInitialized = false)
+        : BaseType((*pMasterGeometry)(), &(pMasterGeometry->msGeometryData)),
```

Comment:
@pooyan-dadvand do you see a possibility to pass the GeometryData of the master to the coupling geometry?

Reply:
I would add a `GeometryData const& GetGeometryData() const` method to base class